File: go/sdk/canvus/color.go
package canvus

import (
	"fmt"
	"regexp"
	"strings"
)

// Canvus API color helpers. All colors are 8-character uppercase hex strings
// in RRGGBBAA format (alpha last; 00 transparent, FF opaque).

var colorRGBAPattern = regexp.MustCompile(`^[0-9A-F]{8}$`)

// ValidateColor returns nil if the input matches the RRGGBBAA uppercase format.
func ValidateColor(color string) error {
	if len(color) != 8 {
		return fmt.Errorf("color must be exactly 8 characters (RRGGBBAA), got %d", len(color))
	}
	if !colorRGBAPattern.MatchString(color) {
		return fmt.Errorf("color must be uppercase hex RRGGBBAA format, got %q", color)
	}
	return nil
}

// NormalizeColor converts an input color string to the Canvus uppercase
// RRGGBBAA format. Accepts: RRGGBBAA (already valid), RRGGBB (adds FF alpha),
// and #-prefixed variants.
func NormalizeColor(color string) (string, error) {
	color = strings.TrimPrefix(color, "#")
	color = strings.ToUpper(color)
	if len(color) == 6 {
		if matched, _ := regexp.MatchString(`^[0-9A-F]{6}$`, color); matched {
			return color + "FF", nil
		}
		return "", fmt.Errorf("invalid 6-character color format: %q", color)
	}
	if err := ValidateColor(color); err != nil {
		return "", err
	}
	return color, nil
}

// ColorToRGBA converts a Canvus color (RRGGBBAA) to separate R, G, B, A bytes.
func ColorToRGBA(color string) (r, g, b, a byte, err error) {
	if err := ValidateColor(color); err != nil {
		return 0, 0, 0, 0, err
	}
	var rr, gg, bb, aa uint
	if _, err := fmt.Sscanf(color, "%02X%02X%02X%02X", &rr, &gg, &bb, &aa); err != nil {
		return 0, 0, 0, 0, fmt.Errorf("failed to parse color %q: %w", color, err)
	}
	return byte(rr), byte(gg), byte(bb), byte(aa), nil
}

// RGBAToColor converts separate RGBA bytes to a Canvus color (RRGGBBAA).
func RGBAToColor(r, g, b, a byte) string {
	return fmt.Sprintf("%02X%02X%02X%02X", r, g, b, a)
}

// ColorToRGB converts a Canvus color (RRGGBBAA) to standard #RRGGBB (no alpha).
func ColorToRGB(color string) (string, error) {
	if err := ValidateColor(color); err != nil {
		return "", err
	}
	return "#" + color[:6], nil
}

// ColorWithAlpha returns a new color string with the alpha channel replaced.
func ColorWithAlpha(color string, alpha byte) (string, error) {
	if err := ValidateColor(color); err != nil {
		return "", err
	}
	return color[:6] + fmt.Sprintf("%02X", alpha), nil
}

// Common opaque colors for convenience.
const (
	ColorBlack       = "000000FF"
	ColorWhite       = "FFFFFFFF"
	ColorRed         = "FF0000FF"
	ColorGreen       = "00FF00FF"
	ColorBlue        = "0000FFFF"
	ColorYellow      = "FFFF00FF"
	ColorCyan        = "00FFFFFF"
	ColorMagenta     = "FF00FFFF"
	ColorGray        = "808080FF"
	ColorLightGray   = "D3D3D3FF"
	ColorDarkGray    = "404040FF"
	ColorTransparent = "00000000"
)

File: typescript/sdk/src/extras/color.ts
// Phase 4b §4.3 #20: color utilities.
//
// Mirrors Go's `color.go`. All Canvus colors are 8-character uppercase hex
// strings in RRGGBBAA format (alpha last; 00 transparent, FF opaque).

const RGBA_RE = /^[0-9A-F]{8}$/;
const RGB_RE = /^[0-9A-F]{6}$/;

/** Throws if `color` is not a valid uppercase RRGGBBAA string. */
export function validateColor(color: string): void {
  if (color.length !== 8) {
    throw new Error(`color must be exactly 8 characters (RRGGBBAA), got ${color.length.toString()}`);
  }
  if (!RGBA_RE.test(color)) {
    throw new Error(`color must be uppercase hex RRGGBBAA format, got ${JSON.stringify(color)}`);
  }
}

/**
 * Normalise an input colour to Canvus uppercase RRGGBBAA.
 *
 * Accepts: `RRGGBBAA`, `RRGGBB` (alpha defaults to `FF`), or either with
 * a leading `#`. Throws on anything else.
 */
export function normalizeColor(input: string): string {
  let color = input.startsWith("#") ? input.slice(1) : input;
  color = color.toUpperCase();
  if (color.length === 6) {
    if (!RGB_RE.test(color)) {
      throw new Error(`invalid 6-character color format: ${JSON.stringify(input)}`);
    }
    return `${color}FF`;
  }
  validateColor(color);
  return color;
}

/** Convert RRGGBBAA → `[r, g, b, a]` (each 0–255). */
export function colorToRgba(color: string): readonly [number, number, number, number] {
  validateColor(color);
  return [
    Number.parseInt(color.slice(0, 2), 16),
    Number.parseInt(color.slice(2, 4), 16),
    Number.parseInt(color.slice(4, 6), 16),
    Number.parseInt(color.slice(6, 8), 16),
  ];
}

/** Compose RRGGBBAA from 0–255 components. */
export function rgbaToColor(r: number, g: number, b: number, a: number): string {
  return [r, g, b, a]
    .map((n) => clamp(n).toString(16).padStart(2, "0").toUpperCase())
    .join("");
}

/** Convert RRGGBBAA → `#RRGGBB` (alpha dropped). */
export function colorToRgb(color: string): string {
  validateColor(color);
  return `#${color.slice(0, 6)}`;
}

/** Return a copy of `color` with its alpha channel replaced. */
export function colorWithAlpha(color: string, alpha: number): string {
  validateColor(color);
  return color.slice(0, 6) + clamp(alpha).toString(16).padStart(2, "0").toUpperCase();
}

function clamp(n: number): number {
  if (!Number.isFinite(n)) return 0;
  return Math.max(0, Math.min(255, Math.round(n)));
}

/** Common opaque colors for convenience. */
export const ColorBlack = "000000FF";
export const ColorWhite = "FFFFFFFF";
export const ColorRed = "FF0000FF";
export const ColorGreen = "00FF00FF";
export const ColorBlue = "0000FFFF";
export const ColorYellow = "FFFF00FF";
export const ColorCyan = "00FFFFFF";
export const ColorMagenta = "FF00FFFF";
export const ColorGray = "808080FF";
export const ColorLightGray = "D3D3D3FF";
export const ColorDarkGray = "404040FF";
export const ColorTransparent = "00000000";
